tests/src/core/testqgsrelation.cpp

```diff
@@ -0,0 +1,112 @@
+/***************************************************************************
+     testqgssettings.cpp
+     --------------------------------------
+    Date                 : 8.03.2022
+    Copyright            : (C) 2022 by Alessandro Pasotti
+    Email                : elpaso at itopen dot it
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#include <QObject>
+
+
+#include "qgsproject.h"
+#include "qgsattributeeditorelement.h"
+#include "qgsattributeeditorcontainer.h"
+#include "qgsattributeeditorrelation.h"
+#include "qgsmaplayerstylemanager.h"
+#include "qgstest.h"
+#include "qgsvectorlayer.h"
+
+
+/**
+ * \ingroup UnitTests
+ * This is a unit test for the QgsRelation changing style
+ */
+class TestQgsRelation: public QObject
```

Comment:
is this change related? Or an accidental commit?

Reply:
It is related: this is a C++ implementation of the original test_qgsrelation.py without the reload changes that I had to introduce in the test.

Long story short: after the introduction of QgsOptionalExpression mCollapsedExpression in https://github.com/qgis/QGIS/pull/47671/files#diff-9a0282d186d8598b3a272358b7517b17b6350908ca79340accd62c5a86852b7dR217 the test_qgsrelation.py started failing, the change is apparently unrelated and cannot be reproduced when the test is run through the QGIS console so I've been running debugging session for hours and decided that it's a SIP issue somewhere that makes the test fail and it is not an issue in real use cases, to add further protection I have re-implemented the original test in C++ without the modifications that I had to make to the python test to make it pass.